### ISSUE TITLE
Feature Request - Extract and wrap with a single child widget

### ISSUE THREAD

**Issue by orestesgaolin** [dart-lang/sdk#52567]
**Is your feature request related to a problem? Please describe.**
Often I need to extract a bigger piece of widget tree to a widget that exposes a single `child` property. This may not be entire subtree but some part of it e.g. from `A->B->C->D` I want to get to `A->B'->D`. To achieve that I create a new widget with `child` property and copy and paste part of the tree that needs to wrap this child. For example starting with following:

```dart
import 'package:flutter/material.dart';

class ExampleParent extends StatelessWidget {
  const ExampleParent({super.key});

  @override
  Widget build(BuildContext context) {
    return const SizedBox(
      width: 400,
      height: 200,
      child: ColoredBox(
        color: Color(0xFF00AA00),
        child: Text('Example'),
      ),
    );
  }
}

```

I end up with this structure:

```dart
import 'package:flutter/material.dart';

class ExampleParent extends StatelessWidget {
  const ExampleParent({super.key});

  @override
  Widget build(BuildContext context) {
    return const SizedColoredBox(
      child: Text('Example'),
    );
  }
}

class SizedColoredBox extends StatelessWidget {
  const SizedColoredBox({super.key, required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 400,
      height: 200,
      child: ColoredBox(
        color: const Color(0xFF00AA00),
        child: child,
      ),
    );
  }
}

```

**Describe the solution you'd like**
I would like to have ability to wrap any widget with a single child wrapper widget similarly as I can extract part of subtree to a new widget.

![CleanShot 2023-05-30 at 16 02 05](https://github.com/Dart-Code/Dart-Code/assets/16854239/84159774-2687-4f21-b373-8755b81dedd2)

**Describe alternatives you've considered**

- Doing this manually
- Extracting entire subtree and reverting this change partially

**Additional context**

Copied from https://github.com/Dart-Code/Dart-Code/issues/4566 as requested by @johnpryan

**Comment by scheglov** [comment]
About specifying how much to extract. For `Extract Local` refactoring we return to the client source ranges of containing expressions, so that IDE can ask the user. I think it works quite conveniently in IntelliJ.
<img width="604" alt="image" src="https://github.com/dart-lang/sdk/assets/384794/a1bba555-dfe2-4591-b012-16f7e320f8e2">
It even highlights visually which portion it will extract.
<img width="608" alt="image" src="https://github.com/dart-lang/sdk/assets/384794/d7c673d8-5fd8-4eda-801b-931ab213e6f2">
We could do something like this here as well.

**Comment by bwilkerson** [comment]
That's an interesting idea. I'm guessing that in the initial example that might look something like the following list:
```
SizedBox
SizedBox, ColoredBox
SizedBox, ColoredBox, Text
```
I'm also guessing that we'd stop before any widget with multiple widgets, because the number of combinations could be large and the options could be confusing.

**Comment by scheglov** [comment]
What combinations? My understanding of the request is that we start at a child widget, and while keeping it where it is, extract a few layers of widgets around it. And if these layers reference other widgets, so let is be, these widgets will go into the extracted widget. And whatever variables these layers reference, will become formal constructor parameters of the new widget.

**Comment by bwilkerson** [comment]
I was assuming that the user would select the top-most widget to be extracted, and that we needed to identify the lowest child being included in the extraction.

Starting at the bottom simplifies our task, but I don't know if that's how users would expect to use the feature.

**Comment by orestesgaolin** [comment]
Personally I don't have any preference, so if starting at the bottom makes things easier, then I don't see any reason to oppose it.

**Comment by johnpryan** [comment]
Maybe instead of supporting N layers of refactoring we suggest a few, like:
- "Extract and wrap with new widget"
- "Extract 2 widgets and wrap with new widget"
- "Extract 3 widgets and wrap with new widget"

**Comment by bwilkerson** [comment]
> I think visually expanding selection is better ...

It definitely would be (visual is almost always better than textual), but I don't know how we would highlight multiple widgets with a hole in the middle for the widget that we wouldn't be extracting.

@DanTup Is that something we can do using LSP (both highlighting two disjoint regions and providing multiple ranges that users can move between)?

If we can't show this visually then we'll need text that defines the range of widgets being extracted in a clear and concise way. That includes indicating whether the widgets are from the selection up the tree or from the selection down the tree.

**Comment by scheglov** [comment]
> > I think visually expanding selection is better ...
> 
> It definitely would be (visual is almost always better than textual), but I don't know how we would highlight multiple widgets with a hole in the middle for the widget that we wouldn't be extracting.

It seems to be possible in IntelliJ.
I hacked [IntroduceTargetChooser](https://github.com/JetBrains/intellij-community/blob/1ec80073addae24892741de7501c35a16c260577/platform/lang-impl/src/com/intellij/refactoring/IntroduceTargetChooser.java#L114) to extract `-2` and `+2` range from the middle of the expression range. Of course this cannot be used as is, but at least the platform allows this on lower levels.
<img width="644" alt="image" src="https://github.com/dart-lang/sdk/assets/384794/51801f45-d85b-4cac-a3f7-aea0dcba62fd">


> If we can't show this visually then we'll need text that defines the range of widgets being extracted in a clear and concise way. That includes indicating whether the widgets are from the selection up the tree or from the selection down the tree.

We need these popup items anyway, highlighting is just a nice addition to it.
And if missing for some text editors, this is probably not too bad.

**Comment by DanTup** [comment]
> @DanTup Is that something we can do using LSP (both highlighting two disjoint regions and providing multiple ranges that users can move between)?

I'm not sure if I completely understand exactly the flow. We can't do anything while the user is moving up and down items in the CodeAction list (we have no code executing there). When are are running code, we could highlight code in VS Code by using decorations (this [was the suggestion](https://github.com/microsoft/vscode/issues/45059) when I asked for an API for highlighting code), but it'd have to be done in the extension (there's no LSP standard for this).

I wonder if an option for his would be to allow the user to just create a selection over the widgets to be extracted?

<img width="476" alt="Screenshot 2023-06-01 at 10 44 58" src="https://github.com/dart-lang/sdk/assets/1078012/5e42572e-20ef-4731-8676-c469ab96679d">

The CodeAction text could be written to indicate this in some way ("Extract SizedBox + ColouredBox", "Extract SizedBox (+ 4 widgets) + ColouredBox").

**Comment by DanTup** [comment]
> I'm guessing that on the VS Code side the equivalent would be to have the refactoring return a list of ranges, use the prompt box to present the list, change the highlight range as the user arrows through the list, and then apply the selected set of edits.

Yep, I think we could do that (although it would be VS Code specific - or at least, require some custom client code and LSP requests which in theory other editors could implement similar to the other new refactor work).
